### PR TITLE
repo: support unlisted but publicly accessible repositories

### PR DESCRIPTION
**Pull request by achilleas-k**
This PR adds support for publicly accessible but unlisted (secret) repositories (issue #5733).

## Forms and UI

I'm including screenshots to show how the new checkbox appears in all relevant forms.

### Creation
new-repository
new-fork
new-migration

### Settings
repo-settings

### Settings (forked)
repo-settings-with-fork

### Icon

Note that, as I mentioned in the relevant issue, I included the icon for testing but it should probably be replaced by something more appropriate, like the [eye-closed](https://primer.style/octicons/eye-closed) octicon in newer versions, or anything you might feel is more appropriate..

repo-home
repo-list

---

## Behaviour

Unlisted repositories behave like Private repositories on pages that list repositories.  The difference is that users who know the URL are allowed to view and clone them.  The general idea is that Unlisted repositories should be "secret", in that no user should be able to discover them, but users who know about them can read them.  In the interest of clarity and posterity, below I list how Unlisted repositories behave in all relevant cases.

For the owner of an unlisted repository:
- They appear on the owner's repository list page.
- They appear on the explore page (and search results).

For a collaborator of an unlisted repository:
- They **do not** appear on the owner's repository list page.
- They appear on the explore page (and search results).

For owners of organisations with unlisted repositories:
- They appear on an organisation's repository list page.

For organisation team members with access to unlisted repositories:
- They appear on an organisation's repository list page.

For other users and guest (non-logged-in) users:
- They **do not** appear on the owner's repository list page.
- They **do not** appear on the explore page (and search results).

General:
- They **do not** appear in the owner's public activity (actions).
- Repositories can be created unlisted.
- Migrations can be created unlisted.
- Forking an unlisted repository copies the unlisted property and the property can't be changed in the fork.
- Changing a repository's unlisted status changes the status for all forks of the repository.
- Changing a repository's unlisted status changes the visibility of associated actions.

---

## API

Adding the Unlisted property to the API requires editing the structs in the go-gogs-client, which I'll do later and then reflect the changes here as necessary.

### CODE REVIEW

**Comment by unknwon** [comment]
Thanks for the PR!

image

This seems much more than needed?

**Comment by achilleas-k** [comment]
> image

That's mostly the last commit that regenerates static assets.

**Comment by unknwon** [comment]
Can you fix it? I can’t view these diffs on GitHub UI. (What I meant was the numbers look wrong)

**Comment by unknwon** [comment]
Please refer to https://github.com/gogs/gogs/blob/master/docs/dev/local_development.md and make sure you’re using the correct go-bindata. https://github.com/kevinburke/go-bindata

**Comment by achilleas-k** [comment]
Hm, it's possible I did a full regen while switching branches.  I'll clean it up and rebase.

**Comment by achilleas-k** [comment]
Regenerated only required `_gen` files with the proper go-bindata.  Sorry about that.

**Comment by unknwon** [comment]
Thanks for the follow up! I'll have this review when I have more bandwidth :)

**Comment by TheRealJP** [comment]
When will this be live? Looks promising, just what a lot of people are looking for! 

Thanks for adding this @achilleas-k

**Comment by unknwon** [comment]
> When will this be live? Looks promising, just what a lot of people are looking for!
> 
> Thanks for adding this @achilleas-k

@TheRealJP It is :) Can't guarantee when though, but just added to my current 0.13 pipeline.

**Comment by achilleas-k** [comment]
Rebasing on main now to keep the PR up to date.

File: internal/db/repo.go [hunk 1]
```diff
@@ -173,8 +173,9 @@ type Repository struct {
 	NumOpenMilestones   int `xorm:"-" gorm:"-" json:"-"`
 	NumTags             int `xorm:"-" gorm:"-" json:"-"`
 
-	IsPrivate bool
-	IsBare    bool
+	IsPrivate  bool
+	IsUnlisted bool
```
Comment (unknwon) [inline internal/db/repo.go:177]:
```suggestion
	// TODO: When migrate to GORM, make sure to do a loose migration with `HasColumn` and `AddColumn`,
	// see docs in https://gorm.io/docs/migration.html.
	IsUnlisted bool
```

**Review by unknwon — APPROVED** [review]
Sorry that it's been a while! The code generally LGTM, just one little thing that we probably want to find a better icon later 😁 But not a blocker to me.

**Comment by unknwon** [comment]
Thanks again and congrats!

**Comment by sigal-raab** [comment]
I am interested in using the unlisted option. However, I don't see it when I access GitHub. Is this option open for everyone?

**Pull request merged**